### PR TITLE
[Core] Support back pressure for actor tasks.

### LINKED ISSUE #18703: [Feature] DirectCall task backpressure required
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Currently, there is no backpressure or flow control on direct call. When transmitting large amounts of data using direct call, the memory usage will rise until oom, if the producer continues to send data without restraint.

Task-based backpressure has been supported in AntGroup for some time. I propose this feature here: https://docs.google.com/document/d/1OGkiOJ-mIxmgNNigO9E7XTCZxzXaj-36u1hQQnAjZgg/edit?usp=sharing

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by MissiontoMars**
## Why are these changes needed?
Resubmit the PR https://github.com/ray-project/ray/pull/19936

I've figure out that the test case `//rllib:tests/test_gpus::test_gpus_in_local_mode` failed due to deadlock in local mode.
In local mode, if the user code submits another task during the executing of current task, the `CoreWorker::actor_task_mutex_` may cause deadlock.
The solution is quite simple, release the lock before executing task in local mode.

In the commit https://github.com/ray-project/ray/pull/20894/commits/7c2f61c76cea3e51a36e36ef06ccaa0b700ddb36:
1. Release the lock in local mode to fix the bug. @scv119 
2. `test_local_mode_deadlock` added to cover the case. @rkooo567 
3. Left a trivial change in `rllib/tests/test_gpus.py` to make the `RAY_CI_RLLIB_DIRECTLY_AFFECTED ` to take effect. @krfricke 
cc @ericl 

## Related issue number

close https://github.com/ray-project/ray/issues/18703

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scv119 — APPROVED** [review]

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1798,6 +1798,11 @@ std::optional<std::vector<rpc::ObjectReference>> CoreWorker::SubmitActorTask(
   RAY_LOG(DEBUG) << "Submitting actor task " << task_spec.DebugString();
   std::vector<rpc::ObjectReference> returned_refs;
   if (options_.is_local_mode) {
+    /// NOTE: The lock should be released in local mode. The user code may
+    /// submit another task when executing the current task locally, which
+    /// cause deadlock. The code call chain is:
+    /// SubmitActorTask -> python user code -> actor.xx.remote() -> SubmitActorTask
+    lock.Release();
```
Comment (scv119) [inline src/ray/core_worker/core_worker.cc:1805]: nice catch.

**Comment by MissiontoMars** [comment]
@scv119 Could you help rerun the buildkite pipeline? The failed tests may be flaky tests.

**Comment by scv119** [comment]
just kicked them off

**Comment by MissiontoMars** [comment]
@ericl The `//rllib:tests/test_gpus` passed. Now I think we can merge this PR.

**Comment by scv119** [comment]
will merge once all tests pass!

**Review by krfricke — APPROVED** [review]
Looks good!

**Review by ericl — APPROVED** [review]

**Pull request merged**